### ISSUE TITLE
pop3 imports which appear in __all__ are not in the pop3 API

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6761]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)| @staceysern reported|
|-|-|
|Trac ID|trac#6761|
|Type|defect|
|Created|2013-09-23 22:07:31Z|
|Branch|https://github.com/twisted/twisted/tree/pop3-imports-6761|

mail.pop3 imports a number of classes from mail.pop3client.  These classes are included in mail.pop.__all__ and they are meant to be accessed through the mail.pop3 module, not mail.pop3client.  However, these classes appear in the API in the mail.pop3client module, not in the mail.pop3 module.

The classes are: 
 * POP3Client (as AdvancedPOP3Client)
 * POP3ClientError
 * InsecureAuthenticationDisallowed
 * ServerErrorResponse
 * LineTooLong

There are two other errors which are defined in pop3client but are not imported into pop3 but probably should be.  They are TLSError and TLSNotSupportedError.

<details><summary>Searchable metadata</summary>

```
trac-id__6761 6761
type__defect defect
reporter__shira shira
priority__normal normal
milestone__ 
branch__branches_pop3_imports_6761 branches/pop3-imports-6761
branch_author__shira shira
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1379974051000000 1379974051000000
changetime__1383483627000000 1383483627000000
version__None None
owner__lvh lvh

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)|@staceysern set owner to @staceysern|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_2"></a>|@staceysern commented|
|-|-|

(In [40083]) Branching to 'pop3-imports-6761'

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_3"></a>|@staceysern commented|
|-|-|

(In [40084]) Initial commit

refs [#6761](https://github.com/twisted/twisted/issues/6761)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_4"></a>|@staceysern commented|
|-|-|

(In [40085]) Add topfile and remove trailing whitespace

refs [#6761](https://github.com/twisted/twisted/issues/6761)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_5"></a>|@staceysern removed owner|
|-|-|

This builds with api documentation errors for three methods on twisted.mail.pop3.AdvancedPOP3Client which is actually twisted.mail.pop3client.POP3Client.  The docstrings in pop3client.py are being changed under ticket [#6653](https://github.com/twisted/twisted/issues/6653) so I am not going to fix the problem here.

twisted.mail.pop3client.__all__ has been changed to the empty list to allow the classes to be included in the twisted.mail.pop3 API.  That means that the behavior of "from twisted.mail.pop3client import *" will change.  However, the docstring for the twisted.mail.pop3client module explicitly says not to use the module directly so hopefully this won't be a problem.

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)|@lvh set owner to @lvh|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_7"></a>|@lvh commented|
|-|-|

Ran buildbots here: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/pop3-imports-6761

There are some failures:
 * https://buildbot.twistedmatrix.com/builders/pyflakes/builds/1168/steps/pyflakes/logs/new%20pyflakes%20errors
 * https://buildbot.twistedmatrix.com/builders/documentation/builds/4100/steps/api-documentation/logs/new%20pydoctor%20errors

This may also be a place to look at an already-existing pyflakes failure:

```
twisted/mail/pop3.py:31: 'twisted' imported but unused
```

Please address and resubmit :)

I'm not sure about the CompatibilityPolicy here. ```pop3client``` doesn't have an underscore, so it's public API, so according to the letter of the document we should have a ```DeprecationWarning```. I'm inclined to invoke "Procedure for Exceptions to this Policy" since it only affects star imports for real code; ```__all__``` is intended as documentation of exported API, and this was wrong. I'd like another opinion on this.

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_8"></a>|@lvh set owner to @staceysern|
|-|-|

Oops, forgot to reassign.

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_9"></a>|@staceysern commented|
|-|-|

(In [40563]) Remove unused import

refs [#6761](https://github.com/twisted/twisted/issues/6761)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_10"></a>|@staceysern commented|
|-|-|

(In [40564]) Remove another unused import

refs [#6761](https://github.com/twisted/twisted/issues/6761)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_11"></a>|@staceysern removed owner|
|-|-|

Although not directly related to this ticket, I removed two unused imports from pop3.py referred to by the pyflakes error in comment 7.  The new pyflake errors reflect a problem with pydoctor and I filed a bug report https://bugs.launchpad.net/pydoctor/+bug/1247519.  The api documentation errors are preexisting but previously appeared for pop3.py and now they appear for pop3client.py.  The docstrings that they occur in are being fixed under ticket [#6653](https://github.com/twisted/twisted/issues/6653).

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_12"></a>|@lvh commented|
|-|-|

I'm a little confused why ```pydoctor``` is related to a ```pyflakes``` bug? I thought the two projects were orthogonal, ```pydoctor``` builds the API docs and pyflakes checks the source code for obvious problems. I would've guessed pyflakes is at fault here?

I could see how this:

```
twisted.mail.pop3.AdvancedPOP3Client.listSize: invalid ref to consumer
twisted.mail.pop3.AdvancedPOP3Client.listUID: invalid ref to consumer
twisted.mail.pop3.AdvancedPOP3Client.retrieve: invalid ref to consumer
```

is a set of ```pydoctor``` bugs, but I'm guessing that those are the docstring deficiencies that you're fixing in 6653 :)

If you unconfuse me on these two points, I'll chase down someone to cast a verdict on whether or not we get to bend the compatibility policy here.

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_13"></a>|@staceysern commented|
|-|-|

Comment 7 mentions three different error: 

 1. The API documentation errors (the ones shown in comment 9).  They are being fixed in [#6653](https://github.com/twisted/twisted/issues/6653).
 2.  The already existing pyflakes error ('twisted' imported but not used).  I fixed that as part of this ticket.
 3.  Other pyflakes errors which show up in the build (https://buildbot.twistedmatrix.com/builders/pyflakes/builds/1168/steps/pyflakes/logs/new%20pyflakes%20errors).  I think that pydoctor incorrectly flags those as errors.

Hope that clears it up.

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_14"></a>|@lvh commented|
|-|-|

Yep, almost: I'm just wondering why you believe that's a pydoctor bug, instead of a pyflakes one. (I think it's a pyflakes bug.)

I'm happy to merge with two outstanding things:

 1. Someone else confirming this doesn't break the compatibility policy
 2. Being convinced that a pyflakes (or pydoctor?) bug is filed somewhere else so that it will eventually get fixed.

I've asked in #twisted-dev for (1) but it being Sunday and all a reply may take a while :)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_15"></a>|@staceysern commented|
|-|-|

I missed the point.  You are right that it is a pyflakes bug not a pydoctor bug.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_16"></a>|@itamarst commented|
|-|-|

There was a module docstring telling people not to use `pop3client`, so this doesn't brought to you.

(This comment brought to you by Pragmatism Weekend™).

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_17"></a>|@lvh set owner to @lvh|
|-|-|

Okay, awesome; I'll merge it then :)

shira, could you please:

 1. Close or comment on the pydoctor bug saying that this is invalid because it's a bug in pyflakes
 2. Open a bug in pyflakes with the same content as the pydoctor bug and link to it here for posterity

Thank you again for your (many) contributions!

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_18"></a>|@lvh set status to `closed`|
|-|-|

(In [40568]) Merge pop3-imports-6761: Correct twisted.mail.pop3's __all__

Author: shira
Reviewer: lvh
Fixes: [#6761](https://github.com/twisted/twisted/issues/6761)

Several POP3 client related classes were documented as being part of the API of the wrong module.

This potentially breaks backwards compatibility for star imports of a module that explicitly told you not to use those classes from there, which was deemed an acceptable bending of the compatibility policy (by lvh and itamar).

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_19"></a>|@staceysern commented|
|-|-|

The pyflakes bug report is https://bugs.launchpad.net/pyflakes/+bug/1247578.
